### PR TITLE
Finishing adressing the PR feedback from Darrel's PR, and fixing an issue

### PR DESCRIPTION
**Pull request by jmprieur**
- Addresses the PR feedback
- Fix one error in the BaseRequestExtension (the type used for the key was wrong)
- Refactored the code that was duplicated in BaseRequestExtension, added XML comments;
- Added two pages to the Web app calls Graph dev app to show how to exercise the new options (`.WithScopes` and `.WithAppOnly`) with links from index.html

@darrelmiller @baywet FYI

@jennyf19 : I tested all the graph dev apps. You might want to try out the Web app calls graph. It's pretty cool :)

### CODE REVIEW

File: src/Microsoft.Identity.Web.MicrosoftGraph/BaseRequestExtensions.cs [hunk 1]
```diff
@@ -15,29 +16,34 @@ public static class BaseRequestExtensions
         /// </summary>
         /// <param name="baseRequest">The <see cref="IBaseRequest"/>.</param>
         /// <param name="scopes">Microsoft graph scopes used to authenticate this request.</param>
-        public static T WithScopes<T>(this T baseRequest, string[] scopes) where T : IBaseRequest
+        public static T WithScopes<T>(this T baseRequest, params string[] scopes) where T : IBaseRequest
```
Comment (jmprieur) [inline src/Microsoft.Identity.Web.MicrosoftGraph/BaseRequestExtensions.cs:19]: Better developer experience.

File: src/Microsoft.Identity.Web.MicrosoftGraph/BaseRequestExtensions.cs [hunk 2]
```diff
@@ -15,29 +16,34 @@ public static class BaseRequestExtensions
...
-            AuthenticationHandlerOption authHandlerOptions = baseRequest.MiddlewareOptions[authHandlerOptionKey] as AuthenticationHandlerOption ?? new AuthenticationHandlerOption();
-            TokenAcquisitionAuthenticationProviderOption msalAuthProviderOption = authHandlerOptions?.AuthenticationProviderOption as TokenAcquisitionAuthenticationProviderOption ?? new TokenAcquisitionAuthenticationProviderOption();
-
-            msalAuthProviderOption.Scopes = scopes;
-
-            authHandlerOptions.AuthenticationProviderOption = msalAuthProviderOption;
-            baseRequest.MiddlewareOptions[authHandlerOptionKey] = authHandlerOptions;
-
-            return baseRequest;
+            return SetParameter(baseRequest, options => options.Scopes = scopes);
```
Comment (jmprieur) [inline src/Microsoft.Identity.Web.MicrosoftGraph/BaseRequestExtensions.cs:21]: Factorizing the duplicated code

File: src/Microsoft.Identity.Web.MicrosoftGraph/BaseRequestExtensions.cs [hunk 3]
```diff
@@ -15,29 +16,34 @@ public static class BaseRequestExtensions
...
+        {
+            string authHandlerOptionKey = typeof(AuthenticationHandlerOption).ToString();
             AuthenticationHandlerOption authHandlerOptions = baseRequest.MiddlewareOptions[authHandlerOptionKey] as AuthenticationHandlerOption ?? new AuthenticationHandlerOption();
             TokenAcquisitionAuthenticationProviderOption msalAuthProviderOption = authHandlerOptions?.AuthenticationProviderOption as TokenAcquisitionAuthenticationProviderOption ?? new TokenAcquisitionAuthenticationProviderOption();
 
-            msalAuthProviderOption.AppOnly = appOnly;
+            action(msalAuthProviderOption);
 
+#pragma warning disable CS8602 // Dereference of a possibly null reference. The Graph SDK ensures it exists
             authHandlerOptions.AuthenticationProviderOption = msalAuthProviderOption;
```
Comment (jmprieur) [inline src/Microsoft.Identity.Web.MicrosoftGraph/BaseRequestExtensions.cs:45]: The Graph SDK does add the authHandlerOptions in the properties of the GraphContext

File: tests/WebAppCallsMicrosoftGraph/Pages/AllApps.cshtml.cs [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+            _graphServiceClient = graphServiceClient;
+        }
+
+        public int NumberOfApps { get; private set; }
+
+        public async Task OnGet()
+        {
+            var messages = await _graphServiceClient.Applications
+                .Request()
+                .WithAppOnly()
```
Comment (jmprieur) [inline tests/WebAppCallsMicrosoftGraph/Pages/AllApps.cshtml.cs:32]: Uses the app only permissions

File: tests/WebAppCallsMicrosoftGraph/Pages/AllUsers.cshtml.cs [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+            _graphServiceClient = graphServiceClient;
+        }
+
+        public int NumberOfUsers { get; private set; }
+
+        public async Task OnGet()
+        {
+            var messages = await _graphServiceClient.Users
+                .Request()
+                .WithScopes("User.Read.All")
```
Comment (jmprieur) [inline tests/WebAppCallsMicrosoftGraph/Pages/AllUsers.cshtml.cs:30]: Uses a specific scopes

File: src/Microsoft.Identity.Web.MicrosoftGraph/BaseRequestExtensions.cs [hunk 6]
```diff
@@ -15,29 +16,34 @@ public static class BaseRequestExtensions
         /// </summary>
         /// <param name="baseRequest">The <see cref="IBaseRequest"/>.</param>
         /// <param name="scopes">Microsoft graph scopes used to authenticate this request.</param>
-        public static T WithScopes<T>(this T baseRequest, string[] scopes) where T : IBaseRequest
+        public static T WithScopes<T>(this T baseRequest, params string[] scopes) where T : IBaseRequest
         {
-            string authHandlerOptionKey = typeof(TokenAcquisitionAuthenticationProviderOption).ToString();
```
Comment (jmprieur) [inline src/Microsoft.Identity.Web.MicrosoftGraph/BaseRequestExtensions.cs:20]: @jennyf19 : this was the issue i was strugging with. 
Shouldn't have been TokenAcquisitionAuthenticationProviderOption, but AuthenticationHandlerOption

cc: @darrelmiller FYI

Comment (jennyf19) [reply src/Microsoft.Identity.Web.MicrosoftGraph/BaseRequestExtensions.cs:20]: i see.

---
In reply to: [509692341](https://github.com/AzureAD/microsoft-identity-web/pull/709#discussion_r509692341) [](ancestors = 509692341)

**Comment by jennyf19** [comment]
[PR](https://github.com/AzureAD/microsoft-identity-web/pull/711)

**Review by jennyf19 — APPROVED** [review]
:shipit: - have a [PR](https://github.com/AzureAD/microsoft-identity-web/pull/711) which some changes to include first. nice work on this.

**Pull request merged**